linkerd/app/core/src/errors/respond.rs

```diff
@@ -76,76 +121,138 @@ impl<ReqB, RspB: Default + hyper::body::HttpBody>
-                    .client
-                    .as_ref()
-                    .map(|ClientHandle { ref addr, .. }| *addr)
-                    .unwrap_or_else(|| {
-                        debug!("Missing client address");
-                        ([0, 0, 0, 0], 0).into()
-                    });
-                warn!(client.addr = %addr, "Failed to proxy request: {}", error);
-
-                if self.version == http::Version::HTTP_2 {
-                    if let Some(reset) = error.h2_reason() {
-                        debug!(%reset, "Propagating HTTP2 reset");
-                        return Err(error);
-                    }
-                }
+impl<B, R> respond::Respond<http::Response<B>, Error> for Respond<R>
+where
+    B: Default + hyper::body::HttpBody,
+    R: HttpRescue<Error> + Clone,
+{
+    type Response = http::Response<ResponseBody<R, B>>;
 
-                // Gracefully teardown the server-side connection.
-                if let Some(ClientHandle { ref close, .. }) = self.client.as_ref() {
-                    debug!("Closing server-side connection");
-                    close.close();
-                }
+    fn respond(&self, res: Result<http::Response<B>>) -> Result<Self::Response> {
+        let error = match res {
+            Ok(rsp) => {
+                return Ok(rsp.map(|b| match self {
+                    Respond {
+                        is_grpc: true,
+                        rescue,
+                        ..
+                    } => ResponseBody::GrpcRescue {
+                        inner: b,
+                        trailers: None,
+                        rescue: rescue.clone(),
+                    },
+                    _ => ResponseBody::Passthru(b),
+                }));
+            }
+            Err(error) => error,
+        };
 
-                // Set the l5d error header on all responses.
-                let mut builder = http::Response::builder();
-                builder = set_l5d_proxy_error_header(builder, &*error);
-
-                if self.is_grpc {
-                    let mut rsp = builder
-                        .version(http::Version::HTTP_2)
-                        .header(http::header::CONTENT_LENGTH, "0")
-                        .header(http::header::CONTENT_TYPE, GRPC_CONTENT_TYPE)
-                        .body(ResponseBody::default())
-                        .expect("app::errors response is valid");
-                    let code = set_grpc_status(&*error, rsp.headers_mut());
-                    debug!(?code, "Handling error with gRPC status");
-                    return Ok(rsp);
-                }
+        let span = info_span!(
+            "rescue",
+            client.addr = %self.client
+                .as_ref()
+                .map(|ClientHandle { addr, .. }| *addr)
+                .unwrap_or_else(|| {
+                    tracing::debug!("Missing client address");
+                    ([0, 0, 0, 0], 0).into()
+                })
```

Comment:
nit/TIOLI: maybe this should go in a method, so that the span macro can just call `self.client_addr()` or something? not a blocker